### ISSUE TITLE
grpc-dotnet + Blazor sets Safari and Firefox User-Agent to "grpc-dotnet/$assemblyVersion"

### ISSUE THREAD

**Issue by LukasReschke** [dotnet/aspnetcore#29183]
### Describe the bug

When using grpc-dotnet and Blazor together to call a GRPC API, the User-Agent is set to "grpc-dotnet/$assemblyVersion" in Safari and Firefox. In Chrome and Edge it is the actual browser User-Agent.

https://docs.microsoft.com/en-us/aspnet/core/grpc/services?view=aspnetcore-5.0 for example describes how to read the User-Agent:

```cs
public override Task<ExampleResponse> UnaryCall(ExampleRequest request, ServerCallContext context)
{
    var userAgent = context.RequestHeaders.GetValue("user-agent");
    // ...

    return Task.FromResult(new ExampleResponse());
}
```

(Migrated from https://github.com/dotnet/AspNetCore.Docs/issues/21178)

### To Reproduce

Use GRPC with Blazor as described at https://docs.microsoft.com/en-us/aspnet/core/grpc/browser?view=aspnetcore-5.0 and look at the traffic in Safari. The User-Agent will be for example "grpc-dotnet/2.34.0.0".

You can also check out https://our.gatekeeper.page in Safari or Firefox and look at the request to the "WhoAmI" endpoint to see the header being wrong. The code for this can be found at https://github.com/getgatekeeper/server.

The Blazor example from https://github.com/grpc/grpc-dotnet/tree/master/examples/Blazor should also have the same issue.

The issue seems potentially related to the following piece of code in [`Grpc.Net.Client/Internal/GrpcCall.cs`](https://github.com/grpc/grpc-dotnet/blob/13a4bf1b2dea216c055370a8ec89827e0635f75d/src/Grpc.Net.Client/Internal/GrpcCall.cs#L798-L799). I wonder if Chrome/Edge just ignore the header being overwritten whilst Safari and Firefox obeys it: 

```cs
// User agent is optional but recommended.
headers.TryAddWithoutValidation(GrpcProtocolConstants.UserAgentHeader, GrpcProtocolConstants.UserAgentHeaderValue);
```

## Test plan
| Browser  | Sends correct User-Agent | Screenshot |
| ------------- | -------------  | ------------- |
| Safari  | ❌  |![image](https://user-images.githubusercontent.com/878997/104110255-4b048b00-52d6-11eb-91fc-fd6f22942a05.png)| 
| Firefox  | ❌  |![image](https://user-images.githubusercontent.com/878997/104110451-185b9200-52d8-11eb-8e66-6656259e2834.png)| 
| Chrome  | ✅   |![image](https://user-images.githubusercontent.com/878997/104110333-0cbb9b80-52d7-11eb-9b6b-63e69597602e.png)|
| Edge  | ✅   |![image](https://user-images.githubusercontent.com/878997/104110350-2d83f100-52d7-11eb-8b8f-fe18167c4697.png)|

**Comment by JamesNK** [comment]
Thanks for the detailed issue. I'll investigate what the client should do here.

@BrennanConroy Does the SignalR JS client set a user-agent?

**Comment by JamesNK** [comment]
It appears the gRPC-Web JS client sets `x-user-agent`: https://github.com/grpc/grpc-web/blob/2e3e8d2c501c4ddce5406ac24a637003eabae4cf/javascript/net/grpc/web/grpcwebclientbase.js#L323

**Comment by BrennanConroy** [comment]
SignalR sets the "X-SignalR-User-Agent" header.

https://github.com/dotnet/aspnetcore/blob/020145ded2fd7ea4657f220e000d64d41232bec3/src/SignalR/clients/ts/signalr/src/Utils.ts#L201

I believe the reason we don't set "User-Agent" is because it used to be a [forbidden header](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name) but no longer is. Also, Chrome just drops the header? https://bugs.chromium.org/p/chromium/issues/detail?id=571722

**Comment by BrennanConroy** [comment]
Oh, I guess this is about WASM, in the .NET client we set the User Agent
https://github.com/dotnet/aspnetcore/blob/020145ded2fd7ea4657f220e000d64d41232bec3/src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs#L597
But you can provide a User Agent header to overwrite.

**Comment by JamesNK** [comment]
@pranavkm Is there a way for a library to detect whether it is running inside the browser? Like something on `Environment` e.g.

```cs
if (!IsBlazorWasm)
{
    headers.Add("user-agent", "library/1.2");
}
else
{
    headers.Add("x-user-agent", "library/1.2");
}
```

**Comment by BrennanConroy** [comment]
Yes it runs in the browser, you can use `OperatingSystem.IsBrowser()`
https://github.com/dotnet/aspnetcore/blob/020145ded2fd7ea4657f220e000d64d41232bec3/src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs#L155
https://github.com/dotnet/aspnetcore/blob/2ceca7fb89a4021166b32f18612bc490d3146fe2/src/Shared/OperatingSystem.cs

**Comment by JamesNK** [comment]
gRPC fixed. Re-opening for SignalR.
